text/0000-remove-stacks-mixins.md

```diff
@@ -0,0 +1,139 @@
+# Meta
+[meta]: #meta
+- Name: Remove Stacks & Mixins
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167), many others
+
+# Summary
+[summary]: #summary
+
+*NOTE: This proposal is part of a larger initiative to reduce complexity originally outlined in https://github.com/buildpacks/rfcs/pull/167*
+
+This RFC proposes that we remove the "stack" and "mixin" concepts from the project and replace them with existing constructs in the container image ecosystem such as base images, Dockerfiles, and OS packages.
+
+# Motivation
+[motivation]: #motivation
+
+The "stack" and "mixin" concepts add unnecessary complexity to the project and make it difficult for new users and contributors to understand how buildpacks work. Compatibility guarantees that are strongly enforced by the stack contract could be replaced with metadata validations and warnings.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Summary of changes:
+- Replace mixins with a [CycloneDX](https://cyclonedx.org)-formatted list of packages.
+- Replace stack metadata (including stack IDs) with canonical OS metadata.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Base Image Metadata
+
+Instead of a stack ID, runtime and build-time base images are labeled with the following canonicalized metadata:
```

Comment:
Will any or all of this data be provided to buildpacks during build or detect? If so:
1. in what format?
2. will run image metadata be provided somehow, or just build image?

We should provide explicit answers to the above in this RFC.

Thoughts on 1:
It seems like we will have to provide this data so that buildpacks, if they support multiple targets, can download and/install the correct assets. We could theoretically force buildpacks to read `/etc/os-release` for some of this, but that doesn't include architecture and it would be nice to provide everything in a consistent format.

Thoughts on point 2:
Right now there are a couple examples in Paketo Java buildpacks where build logic is conditional on the stack ID:
- [native image buildpack stactically compiles](https://github.com/paketo-buildpacks/native-image/blob/main/native/native_image.go#L72-L74) more libraries on the tiny stack
- [Profile buildpack creates direct process for the tiny stack](https://github.com/paketo-buildpacks/procfile/blob/3a6c8fdc14796db671d3697b916792b468307e5e/procfile/build.go#L57)

Given how many Java buildpacks there are, the fact that I am only aware of two examples suggests this isn't _super_ important. The Procfile example in particular is ugly and mostly just demonstrates how trying to fit that existing Procile interface to a distroless world is a bad idea. The native image example is more interesting though, without this data we would need to statically compile always and forgo the ability to patch CVEs in these libraries with rebase.

It's probably not a dealbreaker, but if we aren't providing run image metadata to buildpacks in any way I think this is a change worth calling out explicitly in the RFC.

cc @dmikusa-pivotal

Reply:
For 1: Operating systems generally provide standardized ways of retrieving this information. E.g., `uname`, `lsb_release`. I'm not convinced that it's necessary to create a new interface for it, but I'm also not opposed to it.

For 2: Given that the run image isn't know until immediately before the build process, it seems reasonable to pass an identifier from a run image label to the buildpacks. Could this be a separate (future) RFC?

Reply:
Let me know if the last set of changes addresses these points.

Reply:
Looks OK to me. I like the `CNB_TARGET_*` env variables, as opposed to needing to fish that out of the image manually.